src/circuit/listener.js

```diff
@@ -20,8 +20,8 @@ module.exports = (libp2p) => {
     const deleted = listeningAddrs.delete(connection.remotePeer.toB58String())
 
     if (deleted) {
-      // TODO push announce multiaddrs update
-      // libp2p.identifyService.pushToPeerStore()
+      // Announce multiaddrs update on listen success
+      libp2p.identifyService.pushToPeerStore()
```

Comment:
This is being called in a couple places, I think we shouldn't be doing this here.

Perhaps instead we should be updating our addresses in the PeerStore? We don't currently store them there, but it has an event system for changes, which we could just listen to for triggering a self record update and identify push. This enables the identify system to subscribe to the changes and manage everything itself, instead of other systems like this one needing to know these specifics.

Reply:
We can do that, but it will be tricky.

To enable this, we need to be able to track transportManager `getAddrs()` changes.  With this, once these addresses change, who is responsible to tracking this (perhaps the transportManager while it tracks the transports addrs) needs to update the addressBook for self addresses (with addressBook.set or addressBook.add).

This will cause the event to be triggered as you mentioned, but it will also cause an issue. So, the `identifyService` would listen on address changes for the self peer id and it would create the new self signed peer record, storing it in the peerStore. But, using `addressBook.consumePeerRecord()` will currently trigger a new `address:change` event and get in a loop. 

Regarding the second point, we can use Book `_setData` emit option and put it false if we had the same multiaddrs as before. But, I think we are complicating this a lot. 
I would go with when a self address change, we just create the new peer record and do `addressBook.consumePeerRecord()`, which triggers the event that identify would listen on and Identify does the Push. We can do the same to add own protocol to the protoBook with `add` on handle and remove on unhandle. With this approach, we would be removing the self peer record creation from Identify. We can go back the my first solution using a Record Manager, or we can figure out a way of making this part of the Address Manager.

I would like to have in the future to support dynamic address announce/noAnnounce changes via an API.  So improving the AddressManager to both be responsible for listening for addresses change in the TransportManager, or to the announce/noAnnounce and create/store the self signed peer record would be a good solution. The first point will be difficult to change though.

Reply:
Let's table this for now, it's not a huge deal since it's all internal and I don't think we need to conflate this prematurely.

Reply:
Actually, I think we need to solve this to avoid the unnecessary change checks in `_getSelfPeerRecord` which is going to run on every identify request.

Reply:
Yeah, checking if is different anytime is not a good call. So this is my proposal:

- transportManager will listen for transport [listening events](https://github.com/libp2p/js-libp2p-interfaces/tree/master/src/transport#api).
- when it triggers, transportManager will emit a `listening` event with `getAddrs()` result. It might be also useful for application layer, for example ipfs could use this for the daemon stdout of addresses listening on.
- AddressManager/IdentifyService tracks this event and is responsible for creating the self peer record and add it to the AddressBook. Self addresses will also be updated.

For not complicating too much now, I think we can keep this on the IdentifyService and move it to the Address Manager when we support dynamic API changes. Both need `start` and `stop` methods for the transportManager event listener. The event listener should only be setup after all the initial listeners so that we do not create signed peer records per listener in startup. This way, we will only track updates and keep the current behaviour of creating if it does not exist in the first identify call. If we went with the AddressManager, we would need to have extra logic for this initial creation, or the identifyService would need to call a `addressManager.createSelfPeerRecord` if there was none in the peerStore.